content/codes/components/enigma.ts

```diff
@@ -6,31 +6,598 @@
+    document.getElementById('rotor3setting')!.innerHTML = positions[2];
+    document.getElementById('rotor2setting')!.innerHTML = positions[1];
+    document.getElementById('rotor1setting')!.innerHTML = positions[0];
+  }
+
+  rotate(i:number, up:boolean) {
+    this.machine.rotate(i, up);
+    this.showRotorPositions();
+    this.svgview.drawPath(null);
+  }
+
+  connectRotorControls() {
+    document.getElementById('rotor1up')!.addEventListener("pointerdown", () => this.rotate(0, true));
+    document.getElementById('rotor2up')!.addEventListener("pointerdown", () => this.rotate(1, true));
+    document.getElementById('rotor3up')!.addEventListener("pointerdown", () => this.rotate(2, true));
+    document.getElementById('rotor1dn')!.addEventListener("pointerdown", () => this.rotate(0, false));
+    document.getElementById('rotor2dn')!.addEventListener("pointerdown", () => this.rotate(1, false));
+    document.getElementById('rotor3dn')!.addEventListener("pointerdown", () => this.rotate(2, false));
+    this.showRotorPositions();
+  }
+	
+  updateDisplay(encodingpath:string[])
+  {
+    if (!encodingpath)
+      return;
+    const encodedletter = getLast(encodingpath);
+    this.showRotorPositions();		
+    document.getElementById('cipherpath')!.innerHTML = encodingpath.join("&#x27f6;");
+    const cipherTextEl = document.getElementById('ciphertext') as HTMLElement;
+    let curtext = cipherTextEl.innerHTML;
+    var l = curtext.replace(/[\s+\\n]/g, "").length;
+    if (l && l % 50 == 0)
+      curtext += "\n";
+    else if (l && l % 5 == 0)
+      curtext += " ";
+    curtext += encodedletter;
+    cipherTextEl.innerHTML = curtext;
   }
 }
+
+class PlugboardView
+{  
+  private readonly plugboard = document.getElementById("plugboard");
+  private readonly enigmaview:HTMLElement;
+  private readonly machine:Machine;
+  private readonly svgview:any;
+  
+  private dragging = false;
+  private mouseStart:number[] | null = null;
+  private startPlug:HTMLElement | null = null;
+  private join = false;
+  private path:SVGPathElement | null = null;
+  private cablecount = 0;
+  private curcable = 0;
+  private cables: { [letter: string]: number} = {};
+  private freecables = [true, true, true, true, true, true, true, true, true, true];
+
+  constructor(enigmaviewid:string, machine:Machine, svgview:any) {
+    this.enigmaview = document.getElementById(enigmaviewid) as HTMLElement;
+    this.machine = machine;
+    this.svgview = svgview;
+    // avoid the I-beam cursor bug when dragging cables
+    this.enigmaview.onmouseover = () =>  { document.onselectstart = function() { return false; } };
+    this.enigmaview.onmouseout = () =>  { document.onselectstart = function() { return true; } };
+
+    this.setupListeners();
+  }
+
+  setupListeners() {
+    const plugboard = this.machine.getPlugboard();
+    document.querySelectorAll<HTMLElement>(".plug").forEach(p => {
+    	p.addEventListener('mousedown', e =>
```

Comment:
We should never bind mouse events directly – let's use pointer events instead, to ensure compatibility with touch devices, or (even better) use the built-in `slide()` function for gesture handling. Documentation [here](https://www.notion.so/mathigon/Engineering-5272e83217004731bca2b63d03e654ea#e9c56135f62445c19c919763c028391f)